FishKolm/src/FishKolm.cpp

```diff
@@ -0,0 +1,388 @@
+{
+  SolverControl solver_control(100000, 1e-6 * system_rhs.l2_norm());
+
+  SolverCG<TrilinosWrappers::MPI::Vector> solver(solver_control);
+  TrilinosWrappers::PreconditionSSOR      preconditioner;
+  preconditioner.initialize(
+    lhs_matrix, TrilinosWrappers::PreconditionSSOR::AdditionalData(1.0));
+
+  solver.solve(lhs_matrix, delta_owned, system_rhs, preconditioner);
+  pcout << "  " << solver_control.last_step() << " CG iterations" << std::endl;
+}
+
+// If having problems with the CG solver, try this alternative method.
+void 
+FISHKOLM::solve_linear_system_2()
+{
+  SolverControl solver_control(5000, 1e-6 * system_rhs.l2_norm());
+  SolverCG<TrilinosWrappers::MPI::Vector> solver(solver_control);
+  TrilinosWrappers::PreconditionAMG preconditioner;
+  TrilinosWrappers::PreconditionAMG::AdditionalData amg_data;
+  preconditioner.initialize(lhs_matrix, amg_data);
+
+  solver.solve(lhs_matrix, delta_owned, system_rhs, preconditioner);
+  pcout << "  " << solver_control.last_step() << " CG iterations" << std::endl;
+}
+
+unsigned int 
+FISHKOLM::solve_newton_opt()
+{
+  const unsigned int n_max_iters = 1000;
+  const double residual_tolerance = 1e-3;
+
+  unsigned int n_iter = 0;
+  double residual_norm = residual_tolerance + 1;
+
+  while (n_iter < n_max_iters && residual_norm > residual_tolerance)
+  {
+    assemble_system();
+    residual_norm = system_rhs.l2_norm();
+
+    pcout << "  Newton iteration " << n_iter << "/" << n_max_iters
+          << " - ||r|| = " << std::scientific << std::setprecision(6)
+          << residual_norm << std::flush;
+
+    // We actually solve the system only if the residual is larger than the
+    // tolerance.
+    if (residual_norm > residual_tolerance)
+    {
+      solve_linear_system();
+      solution_owned += delta_owned;
+      solution = solution_owned;
+    }
+    else
+    {
+      pcout << " < tolerance" << std::endl;
+    }
+
+    ++n_iter;
+  }
+  return n_iter;
+}
+
+void 
+FISHKOLM::solve_newton()
+{
+  const unsigned int n_max_iters = 1000;
+  const double residual_tolerance = 1e-3;
+
+  unsigned int n_iter = 0;
+  double residual_norm = std::numeric_limits<double>::max();
+
+  while (n_iter < n_max_iters && residual_norm > residual_tolerance) {
+    assemble_system();
+    residual_norm = system_rhs.l2_norm();
+
+    pcout << "  Iteration: " << n_iter << ": "
+          << " ||r|| = " << std::scientific << std::setprecision(3)
+          << residual_norm << std::flush;
+    
+    if (residual_norm <= residual_tolerance)
+    {
+      pcout << " < tolerance" << std::endl;
+      break;
+    }
+    
+    pcout << std::endl;
+    solve_linear_system();
+    solution_owned += delta_owned;
+    solution = solution_owned;
+
+    ++n_iter;
+  }
+
+  if (n_iter == n_max_iters && residual_norm > residual_tolerance)
+  {
+    pcout << "  Warning: Newton solver did not converge after "
+          << n_max_iters << " iterations. Final residual: "
+          << residual_norm << std::endl;
+  }
+}
```

Comment:
Similar to the above: what is the difference between these two methods? Does it really make sense to keep both, or is there one that is strictly better than the other, so that the latter can be discarded altogether?